core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java

```diff
@@ -1059,4 +1035,117 @@ public final Object visitGetKeysInGroupCommand(InvocationContext ctx,
+         InternalCacheEntry ice = micv.toInternalCacheEntry(command.getKey());
+         returnValue = ice.getValue();
+
+         // TODO: skip lookup by returning from entry factory directly
+         entryFactory.wrapExternalEntry(ctx, command.getKey(), ice, true, true);
+         cacheEntry = ctx.lookupEntry(command.getKey());
+         cacheEntry.setChanged(true);
+         return apply(ctx, command, null);
+      }
+
+      @Override
+      public Object apply(InvocationContext rCtx, VisitableCommand rCommand, Object rv) throws Throwable {
+         DataWriteCommand cmd = (DataWriteCommand) rCommand;
+         // We don't care about the local value, as we use MATCH_ALWAYS on backup
+         boolean committed = commitSingleEntryIfNewer(cacheEntry, rCtx, cmd);
+         if (committed && !cmd.hasAnyFlag(FlagBitSets.PUT_FOR_STATE_TRANSFER)) {
+            svm.scheduleKeyInvalidation(cmd.getKey(), cacheEntry.getMetadata().version(), cacheEntry.isRemoved());
+         }
+         return returnValue;
+      }
+
+      @Override
+      public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
+         return handleDataWriteCommand(ctx, command);
+      }
+
+      @Override
+      public Object visitRemoveCommand(InvocationContext ctx, RemoveCommand command) throws Throwable {
+         return handleDataWriteCommand(ctx, command);
+      }
+
+      @Override
+      public Object visitReplaceCommand(InvocationContext ctx, ReplaceCommand command) throws Throwable {
+         return handleDataWriteCommand(ctx, command);
+      }
+
+      @Override
+      public Object visitComputeCommand(InvocationContext ctx, ComputeCommand command) throws Throwable {
+         return handleComputeCommand(ctx, command);
+      }
+
+      @Override
+      public Object visitComputeIfAbsentCommand(InvocationContext ctx, ComputeIfAbsentCommand command) throws Throwable {
+         return handleComputeCommand(ctx, command);
+      }
+
+      @Override
+      public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
+         throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public Object visitReadWriteKeyValueCommand(InvocationContext ctx, ReadWriteKeyValueCommand command) throws Throwable {
+         throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public Object visitReadWriteKeyCommand(InvocationContext ctx, ReadWriteKeyCommand command) throws Throwable {
+         throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public Object visitReadWriteManyCommand(InvocationContext ctx, ReadWriteManyCommand command) throws Throwable {
+         throw new UnsupportedOperationException();
+      }
+
+      @Override
+      public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteManyEntriesCommand command) throws Throwable {
+         throw new UnsupportedOperationException();
```

Comment:
My mistake, I saw the method names `handleDataWriteCommand` and `handleComputeCommand` and I assumed that `ComputeCommand` is not a `DataWriteCommand`. Since `handleWritePrimaryResponse` casts the command to a `DataWriteCommand`, I further assumed that there must be another method using a `PrimaryResponseHandler` with a `ComputeCommand`, and that you have a neat separation between the two use cases :)